docs/internals_overview.md

```diff
@@ -0,0 +1,382 @@
+the congestion strategy in lokinet happens as a coincidental side effect in the implementation detail of how link layer frames are transported.
+elevated packet loss causes retransmissions which will delay traffic to the next hop. 
+this behavior is desirable but because it is done at the link layer and not the onion layer, it will affect all paths going across that node.
+this behavior should be reimplemented on an upper level, propagating link layer session info upwards would help inform how and when to do this.
+
+iwp is a homegrown protocol that nothing else speaks, it needs to be replaced.
+DTLS + SCP would be a prime choice as it is what webrtc speaks, this would make a WASM port a possiblity.
+QUIC would be great for durable transmission of messages but the capabilities of it to transmit lossy data is unclear at this time.
+
+### Flow Layer
+
+this layer needs to be redone at a high level in full so that the role it actually has is well defined, the timeout, limits and other constraints of the protocol are well defined, the structure of state machines involved are well defined, the failure cases are well defined, the exceptional failures are well defined and the interaction with the lower layers is well defined.
+there is no well defined anything in the flow layer as nothing in it was designed, it was organically piled together. 
+this layer is responsible for a very large majority of issues in lokinet (likely > 70% of user facing issues).
+
+
+## implementation problems
+
+* this codebase currently comes with 2 ½ kitchen sinks pre-installed.
+* there are not enough strict separation between the concerns of components.
+* there is an over dependance on inheritence in dubious and ambigous ways.
+* there are home grown protocols in use for link layer comms.
+* the platform specific code is not well contained in their own domain.
+* the rpc server.
+* the dht subsystem.
+* there is no internal documentation of any kind.
+
+### kitchen sinks
+
+* `llarp::Router` 
+* the service endpoint amalgum
+
+#### llarp::Router
+
+the contextualized god object is supposed to be `llarp::Context` but well over 90% of the functionality is done in `llarp::Router` and the former provides a psuedo wrapper with a semi-stable C++ api. this sucks, alot. this needs to be split up and refactored.
+
+#### the service endpoint amalgum
+
+this is the 1 ½ kitchen sink, a signle giant component that strattles the `llarp::handlers` and `llarp::service` that is in charge of the flow layer.
+none of the flow layer was designed before hand and this code has had 2 or 3 partial failed refactors laying inside of it.
+this code is the result of an additive organic dev process with zero design, foresight or rigor. 
+it is a jumbled clusterfuck of obtuse partially completed components that merge into itself, fractally, with ambiguous names of functions, types and methods, with no useful documentation of any kind. (minus occational comments peppered with explitives).
```

Comment:
Please ditch the hyperbole, keep this doc professional.

Reply:
yup, this was cherrypicked from a local branch